### PR TITLE
New design

### PR DESCRIPTION
**Pull request by OliverSchw**

### CODE REVIEW

**Comment by hvater** [comment]
IMO everything should be in english. This is just so that everyone can work with this repo.

For example the names of a lot of the notebooks are in german

File: exciting_environments/core_env.py [hunk 1]
```diff
@@ -149,109 +143,101 @@ def _test_rew_func(self, func):
...
+            out = jax.vmap(func, in_axes=(0, 0, self.in_axes_env_properties.action_constraints))(
+                jnp.zeros((self.batch_size, int(len(self.obs_description)))),
+                jnp.zeros((self.batch_size, int(
+                    len(list(vars(self.env_properties.action_constraints).values()))))),
+                self.env_properties.action_constraints
+            )
         except:
             raise Exception(
-                "Reward function should be using obs matrix as only parameter")
+                "Reward function should be using observation matrix, action and self.env_properties.action_constraints as parameters")
```
[the remark this answers is not in the document]
Comment (hvater) [reply exciting_environments/core_env.py:154]: I think this error message could be a bit confusing. There are certainly cases where the function call breaks and the reason is something other than the input arguments. Instead of doing a try -> except statement here, you should just test a couple of different inputs and the output shapes here.

I am also unsure if we need to provide a test for the reward function. We should present a e.g. a header or documentation.

File: exciting_environments/core_env.py [hunk 2]
```diff
@@ -149,109 +143,101 @@ def _test_rew_func(self, func):
...
         try:
             if out.shape != (self.batch_size, 1):
                 raise Exception(
-                    "Reward function should be returning vector in shape (batch_size,1)")
+                    "Reward function should be returning vector in shape (1,)")
         except:
             raise Exception(
-                "Reward function should be returning vector in shape (batch_size,1)")
+                "Reward function should be returning vector in shape (1,)")
```
Comment (hvater) [inline exciting_environments/core_env.py:161]: Similar issue to the problem above. There could be other cases where the call breaks because of different reasons. I do not think that a try and except formulation is helpful here.

File: exciting_environments/core_env.py [hunk 3]
```diff
@@ -149,109 +143,101 @@ def _test_rew_func(self, func):
...
         Returns:
             Multiple Outputs:
 
-            observation(ndarray(float)): Observation/State Matrix (shape=(batch_size,states)).
-
-            reward(ndarray(float)): Amount of reward received for the last step (shape=(batch_size,1)).
+            observation: The gathered observation.
 
-            terminated(bool): Flag, indicating if Agent has reached the terminal state.
+            reward: Amount of reward received for the last step.
```
Comment (hvater) [inline exciting_environments/core_env.py:179]:
```suggestion
```

File: exciting_environments/core_env.py [hunk 4]
```diff
@@ -149,109 +143,101 @@ def _test_rew_func(self, func):
...
 
-            truncated(ndarray(bool)): Flag, indicating if state has gone out of bounds (shape=(batch_size,states)).
+            terminated: Flag, indicating if Agent has reached the terminal state.
 
-            {}: An empty dictionary for consistency with the OpenAi Gym interface.
+            truncated: Flag, indicating if state has gone out of bounds.
```
Comment (hvater) [inline exciting_environments/core_env.py:182]: Going out of bounds is not the only case where an environment could be truncated. A different case would be an artifical timer that is independent of the actual task.

See [here](https://gymnasium.farama.org/tutorials/gymnasium_basics/handling_time_limits/#truncation) for more information.

File: exciting_environments/core_env.py [hunk 5]
```diff
@@ -149,109 +143,101 @@ def _test_rew_func(self, func):
...
         # ode step
-        states = jax.vmap(self._ode_exp_euler_step)(states, action_norm, self.env_state_normalizer,
-                                                    self.action_normalizer, self.static_params)
+        states = self._ode_solver_step(
+            states, action, env_properties.static_params)
 
         # observation
-        obs = jax.vmap(self.generate_observation)(states)
+        obs = self.generate_observation(
+            states, env_properties.physical_constraints)
```
Comment (hvater) [inline exciting_environments/core_env.py:193]: This is something that I have seen a few times already, so this is not specific to this function call here.

Following the [python google style guide's section on line breaks](https://google.github.io/styleguide/pyguide.html#32-line-length) my preferred way to do line breaks is like this:

```suggestion
        obs = self.generate_observation(
            states, env_properties.physical_constraints
        )
```

```
output = func(a, b, c, foo, bar, very_long_input_so_that_a_new_line_is_necessary,
              states, foo2, bar2)
```

is also absolutely valid. But this way of formatting does not work so well in this specific case.

What you are doing a lot is a bit of a combination of the two, which I do not really like as it is less
clear which brackets belong together IMO.

File: exciting_environments/core_env.py [hunk 6]
```diff
@@ -149,109 +143,101 @@ def _test_rew_func(self, func):
...
-        reward = jax.vmap(self.reward_func)(obs, action_norm).reshape(-1, 1)
+        reward = self.reward_func(
+            obs, action, env_properties.action_constraints)
 
         # bound check
-        truncated = jax.vmap(self.generate_truncated)(states)
-        terminated = jax.vmap(self.generate_terminated)(states, reward)
+        truncated = self.generate_truncated(
+            states, env_properties.physical_constraints)
+        terminated = self.generate_terminated(states, reward)
```
Comment (hvater) [inline exciting_environments/core_env.py:202]: Why does this take the reward?

File: exciting_environments/core_env.py [hunk 7]
```diff
@@ -149,109 +143,101 @@ def _test_rew_func(self, func):
...
+            states: The states from which to calculate states for the next step.
+            action: The action to apply to the environment.
+            env_properties: Contains action/state constraints and static parameters.
 
 
         Returns:
             Multiple Outputs:
 
-            observation(ndarray(float)): Observation/State Matrix (shape=(batch_size,states)).
+            observation: The gathered observations (shape=(batch_size,obs_dim)).
```
Comment (hvater) [inline exciting_environments/core_env.py:220]:
```suggestion
```

File: exciting_environments/core_env.py [hunk 8]
```diff
@@ -262,33 +248,38 @@ def default_reward_func(self, obs, action):
     @partial(jax.jit, static_argnums=0)
     @abstractmethod
     def generate_observation(self, states):
-        """Returns states."""
+        """Returns observation."""
         return states
```
Comment (hvater) [inline exciting_environments/core_env.py:252]: I think this should not return anything
```suggestion
        return
```
```suggestion
        pass
```
```suggestion
        raise NotImplementedError
```

are preferable IMO.

File: exciting_environments/mass_spring_damper/mass_spring_damper_env.py [hunk 9]
```diff
@@ -15,121 +18,179 @@ class MassSpringDamper(core_env.CoreEnvironment):
...
+    ):
         """
         Args:
             batch_size(int): Number of training examples utilized in one iteration. Default: 8
-            d(float): Damping constant. Default: 1
-            k(float): Spring constant. Default: 100
-            m(float): Mass of the oscillating object. Default: 1
-            max_force(float): Maximum force that can be applied to the system as action. Default: 20
-            reward_func(function): Reward function for training. Needs Observation-Matrix and Action as Parameters. Default: None (default_reward_func from class) 
+            physical_constraints(jdc.pytree_dataclass): Constraints of physical states of the environment.
```
Comment (hvater) [inline exciting_environments/mass_spring_damper/mass_spring_damper_env.py:58]: same as for the cartpole. We are expecting a dictionary, but the documentation tells us that it expects a dataclass. I understand where this comes from, but this needs to be acknowledged IMO.

File: exciting_environments/mass_spring_damper/mass_spring_damper_env.py [hunk 10]
```diff
@@ -15,121 +18,179 @@ class MassSpringDamper(core_env.CoreEnvironment):
...
+        opt = None  # self.Optional(something=jnp.zeros(self.batch_size))
+        return self.States(physical_states=phys, PRNGKey=None, optional=opt)
 
     @partial(jax.jit, static_argnums=0)
-    def default_reward_func(self, obs, action):
-        return ((obs[0])**2 + 0.1*(obs[1])**2 + 0.1*(action[0])**2)
+    def default_reward_func(self, obs, action, action_constraints):
+        """Returns reward for one batch."""
+        reward = (obs[0])**2 + 0.1*(obs[1])**2 + 0.1 * \
+            (action[0]/action_constraints.force)**2
```
Comment (hvater) [inline exciting_environments/mass_spring_damper/mass_spring_damper_env.py:154]:
```suggestion
        reward = ((obs[0])**2 + 0.1*(obs[1])**2 
                  + 0.1 * (action[0]/action_constraints.force)**2)
```

Brackets for line continuation. I like to split at the additions and IMO the operator should be the first element in the new line.

File: exciting_environments/mass_spring_damper/mass_spring_damper_env.py [hunk 11]
```diff
@@ -15,121 +18,179 @@ class MassSpringDamper(core_env.CoreEnvironment):
...
-    def generate_truncated(self, states):
-        """Returns states."""
-        return jnp.abs(states) > 1
+    def generate_truncated(self, states, physical_constraints):
+        """Returns truncated information for one batch."""
+        _states = jnp.hstack((
+            states.physical_states.deflection / physical_constraints.deflection,
+            states.physical_states.velocity / physical_constraints.velocity,
+        ))
+        return jnp.abs(_states) > 1
```
Comment (hvater) [inline exciting_environments/mass_spring_damper/mass_spring_damper_env.py:177]: Isn't this the same as 

```suggestion
        _obs = self.generate_observation(states, physical_constraints)
        return jnp.abs(_obs) > 1
```

A bit clearer and code reuse is (almost) always good

File: exciting_environments/mass_spring_damper/mass_spring_damper_env.py [hunk 12]
```diff
@@ -15,121 +18,179 @@ class MassSpringDamper(core_env.CoreEnvironment):
...
         return reward == 0
 
-    @property
-    def obs_description(self):
-        return self.states_description
+    def reset(self, rng: chex.PRNGKey = None, initial_states: jdc.pytree_dataclass = None):
+        """Resets environment to default or passed initial states."""
+        if initial_states is not None:
+            assert tree_structure(self.init_states()) == tree_structure(
+                initial_states), f"initial_states should have the same dataclass structure as self.init_states()"
```
Comment (hvater) [inline exciting_environments/mass_spring_damper/mass_spring_damper_env.py:188]:
```suggestion
            assert tree_structure(self.init_states()) == tree_structure(initial_states), (
                f"initial_states should have the same dataclass structure as self.init_states()"
            )
               
```

File: exciting_environments/pendulum/pendulum_env.py
Comment (hvater) [inline exciting_environments/pendulum/pendulum_env.py:1]: the issues here seem to be the same as in mass_spring_damper_env.py so refer to my comments on that

**Review by hvater — COMMENTED** [review]
Overall, this looks very good to me. Nice work.

I mostly commented small things and formatting. Please look through my comments and tell me what you think.

**Comment by hvater** [comment]
One more thing: We should really get the tests to work before merging.

Seems to be that diffrax is missing from the requirements?

**Comment by hvater** [comment]
> _ No description provided. _

Also a small description of what we have changed would be nice. It is not so important at this point, but it is a good practice to do things thoroughly from the beginning
